CC-8803: Add support for upsert/delete via periodic merge flushes

C0urante opened the pull request:
[Jira](https://confluentinc.atlassian.net/browse/CC-8803)

This implements the changes proposed in the [collaborative design doc](https://docs.google.com/document/d/1p8_rLQqR9GIALIruB3-MjqR8EgYdaEw2rlFF1fxRJf0) for upsert/delete support in the connector.

ncliang commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -343,12 +472,16 @@ public void start(Map<String, String> properties) {
             config.getBoolean(config.BIGQUERY_PARTITION_DECORATOR_CONFIG);
     if (hasGCSBQTask) {
       startGCSToBQLoadTask();
+    } else if (upsertDelete) {
```
I am not sure I am understanding this. It looks like `hasGCSBQTask` is a flag on task0 that is set if `ENABLE_BATCH_CONFIG` is not empty. In which case, a background thread is started to do some batch loading. If both batch loading and upsert delete is enabled, would task0 just get the batch loading background thread and not the merge flush thread?

C0urante replied:
Ah, forgot to add some validation logic for this. Basically, GCS batch loading will be disabled if upsert/delete is enabled and if the user tries to enable both a validation error will be thrown during startup.

ncliang commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java:
```diff
@@ -0,0 +1,336 @@
+        } catch (Throwable t) {
+          logger.warn("Failed on merge flush from intermediate table {} to destination table {}",
+              intermediateTable, destinationTable, t);
+          throw new ConnectException(
```
This exception actually will get swallowed unless you handle the `ExecutionException` thrown on `Future.get`. https://stackoverflow.com/a/39091808/12414477

Looks like the executor implementation used overrides `afterExecute` hook. You can update it to catch both types of exceptions like this - https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html#afterExecute(java.lang.Runnable,%20java.lang.Throwable)

C0urante replied:
Yep, this is exactly what the connector does already :)

https://github.com/wepay/kafka-connect-bigquery/blob/c4d76ea5a1519aae4a42c22e2d3989caee8a787b/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java#L63-L74

I'll pull out the `Runnable` into a separate method to be clearer what's being invoked by the `executor` field, and make the type signature of that field more specific so that it's obvious that it's an instance of `KCBQThreadPoolExecutor`.

ncliang replied:
The part that I'm worried about is this blurb from javadoc:
```
Note: When actions are enclosed in tasks (such as FutureTask) either explicitly or via methods such as submit, these task objects catch and maintain computational exceptions, and so they do not cause abrupt termination, and the internal exceptions are not passed to this method.
```
This does not seem to be handled in `KCBQThreadPoolExecutor:: afterExecute `

C0urante replied:
🤦‍♀️ no kidding. I should use `execute` here instead of `submit`. Good catch, thanks Nigel!

ncliang commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -251,24 +348,32 @@ private BigQuery getBigQuery() {
     if (testBigQuery != null) {
       return testBigQuery;
     }
+    return bigQuery.updateAndGet(bq -> bq != null ? bq : newBigQuery());
```
Are both `BigQuery` and `SchemaManager` objects thread-safe? Looks like they were not being shared across threads before whereas they may be now. Why share them instead of having them in `ThreadLocal`s and still have just one for each thread?

C0urante replied:
I think they both are used concurrently right now since they get given (directly or indirectly) to `TableWriterBuilder` instances when are then each run on their own threads.

There are also some changes here to the `SchemaManager` class that should make help alleviate some performance problems caused by concurrent usage by optimizing away redundant operations. They don't make it more thread-safe since it already appears to be, but they do make it friendlier to concurrent use and help avoid unnecessary DDL requests to BigQuery.

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -330,6 +443,22 @@ public void start(Map<String, String> properties) {
+      String intermediateTableSuffix = String.format("_%s_%d_%s_%d",
+          config.getString(config.INTERMEDIATE_TABLE_SUFFIX_CONFIG),
+          config.getInt(config.TASK_ID_CONFIG),
+          uuid,
```
BigQuery table names can be up to 1024 characters: https://cloud.google.com/bigquery/docs/tables#table_naming

It's possible that this suffix will cause table names to exceed that limit, but not super likely.

ncliang replied:
That should be more than long enough.

ncliang commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:
```diff
@@ -55,13 +70,90 @@ public SchemaManager(
+   * @param table The BigQuery table to create,
+   * @param topic The Kafka topic used to determine the schema.
+   */
+  public void createOrUpdateTable(TableId table, String topic) {
```
Are we guaranteed that tasks on different workers won't try to create/update the same table?

C0urante replied:
No, we aren't. This is mostly just a best-effort optimization to prevent lots of small batches from causing tasks to spam BigQuery with DDL requests. We should catch the case where the table create request fails because it already exists, though, and then respond by attempting a table update instead. Good catch, thanks!

ncliang commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/SchemaManager.java:
```diff
@@ -70,9 +162,27 @@ public SchemaManager(
+        if (e.getCode() == 409) {
+          logger.debug("Failed to create {} as it already exists (possibly created by another task)", table(table));
+          schemaCache.put(table, readTableSchema(table));
+        }
```
Do something with the exception if the error code is not 409. Otherwise it would be silently swallowed.

C0urante replied:
🤦‍♀️ this is why we have code review.

C0urante commented on kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java:
```diff
@@ -362,29 +498,50 @@ private void startGCSToBQLoadTask() {
+      return;
+    }
+    logger.info("Attempting to start upsert/delete load executor");
+    loadExecutor = Executors.newScheduledThreadPool(1);
```
Sort of... the executor here just periodically calls `MergeQueries::mergeFlushAll`, which in turn does little more than make a series of calls to `executor::execute`, so the only purpose of this single-threaded executor is to periodically wake up and then schedule the actual merge flushing tasks on a different executor.

See https://github.com/C0urante/kafka-connect-bigquery/blob/6dda50e4707d5df365ef836c3d4b7648097f010f/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/MergeQueries.java#L111

Is there something wrong with this/do you think it can be implemented better somehow?

ncliang replied:
I'm not sure what will happen to the executor instances once the reference to it is assigned to an instance created in another thread. GC may come along and see no references to the instance and decide to GC it without properly shutting it down. Or, it may chug along in the background and you end up with many independently scheduled `mergeFlushAll` calls, each at fixed rate but slightly offset intervals. They likely would not be `shutdown` properly on JVM shutdown or even task rebalances since the reference has been overwritten.

I think it might be better to create the background thread just once on the worker as a singleton instance and have all tasks schedule on the shared instance. Especially since this thread is not doing anything except schedule the `mergeFlushAll` to be executed on another thread. WDYT?

C0urante replied:
Sorry, which executor are you referring to? The `loadExecutor` and `executor` fields are both created and assigned in the same thread that calls `BigQuerySinkTask::start`, and are both cleaned up during `BigQuerySinkTask::stop`. I'm not sure I understand how GC or threading will come into play here?

Using a shared static instance of the executor doesn't seem like a great idea given that there may be task instances of different connectors running on the same worker, but I also don't fully understand the objections to the current model, so there might be something I'm missing here.

ncliang replied:
My understanding is that `BigQuerySinkTask::start` would be called once for each task, each on its own thread. Since the `loadExecutor` field is not thread local, each thread would get access to the shared reference in memory. So if each `start()` method creates a new instance and assigns it to the field, I am not sure what happens to the previous instances created by other threads. I could be misreading the `Worker` initialization code or misunderstanding the threading model entirely though. LMK if you think I am missing something here.

You are right that multiple connectors sharing the global singleton executor instance could be a concern. Rather than have a global singleton, the existing `gcsLoadExecutor` seems to have the connector detect if it is generating config for task zero and explicitly set a flag on the config to tell the task whether it is task zero and needs to create the executor or not. I don't think this will work to create one per worker as the connector doesn't even know what workers the tasks will be scheduled on. You could maybe initialize the executor to null with an AtomicReference to it like you've done for `bigQuery` and `schemaManager`. You'd have to ref count and do executor shutdown when ref count reaches zero though. A bit more involved than a global singleton and a JVM shutdown hook, but it can be done.

C0urante replied:
The `loadExecutor` field is non-static, so each task creates and uses its own. Every thread a task creates will share that executor, but those will still be completely isolated from other tasks. It's similar with the `bigQuery` and `schemaManager` fields; while not thread-local, those are task-local. (And they're atomic references just to avoid potential races when multiple threads from a single task try to initialize them for the first time.)

C0urante replied:
I think the reason the `loadExecutor` field is isolated to a single task when GCS batch loading is enabled isn't to limit sharing of the `Executor` instance itself, but rather, to ensure that there's only one task at a time triggering batch loading jobs in GCS.

ncliang replied:
We discussed offline. Each task would get a new instance and not share the `executor` field with other tasks, so my concern was void. The only real issue I still see here is slightly wasting threads because of a background thread created per task just to schedule the periodic call to `mergeFlushAll`. But, it will be a pain to try to consolidate that to have 1 executor per worker, so we can just leave as is for now and improve it later if this becomes an issue.

C0urante commented:
Thanks @ncliang; I've addressed all of your comments with changes except the remarks about executor instances; still not totally clear on what the idea is there. Happy to discuss offline if that'd be more efficient, lmk :)

C0urante commented on kcbq-connector/src/test/java/com/wepay/kafka/connect/bigquery/SinkPropertiesFactory.java:
```diff
@@ -19,6 +19,7 @@
 
 
 import com.wepay.kafka.connect.bigquery.config.BigQuerySinkConfig;
+import com.wepay.kafka.connect.bigquery.config.BigQuerySinkTaskConfig;
```
```suggestion
```

C0urante commented:
Thanks @ncliang! I'm working on a branch right now to add modernized integration tests that use the `EmbeddedConnectCluster` instead of the existing Docker-based ITs, which don't support testing different configurations of the connector and instead use a single config for every test. If the quality of these changes looks good enough, do you think we can address integration tests via a separate PR?

ncliang commented:
@C0urante . Yeah, I think that is fine. Maybe just try to run existing IT once to do a sanity check make sure we aren't causing any major regression with this PR. Otherwise, the plan to refactor and add IT in a followup PR should be fine.

C0urante commented:
Just ran the integration test locally and everything passed. In my prototype ITs for upsert/delete I've found a few bugs already but we can address those in the same PR that adds those ITs. Gonna merge now.

Thanks for the review Nigel!

ShahNewazKhan commented:
@C0urante thanks for implementing this feature! While running the connector with `upsert/delete enabled` I have found instances of when the `tmp` tables are not merge flushed. 

Is there a way to trigger a merge flush on the `tmp` tables that are stuck in limbo?

C0urante commented:
@ShahNewazKhan I am no longer working on this project. You may want to try reaching out to its current maintainers. Best of luck!

ShahNewazKhan commented:
> @ShahNewazKhan I am no longer working on this project. You may want to try reaching out to its current maintainers. Best of luck!

Got it, thanks!
